tests/unittests/test_runs/test_simple.py

```diff
@@ -0,0 +1,282 @@
+         "type": "ipv4",
+         "routes": [{"netmask": "255.255.255.255",
+                    "network": "192.168.1.0", "gateway": "10.0.1.1"}]}
+    ],
+    "links": [
+        {"ethernet_mac_address": "fa:16:3e:69:b0:58",
+         "id": "eth0", "mtu": 1500, "type": "phy"},
+        {"ethernet_mac_address": "fa:16:3e:b3:72:30",
+         "id": "vlan1", "type": "vlan", "vlan_id": 602,
+         "vlan_link": "eth0", "vlan_mac_address": "fa:16:3e:b3:72:30"},
+    ]
+}
+
+KNOWN_MACS = {
+    'fa:16:3e:69:b0:58': 'enp0s1',
+    'fa:16:3e:d4:57:ad': 'enp0s2',
+    'fa:16:3e:dd:50:9a': 'foo1',
+    'fa:16:3e:a8:14:69': 'foo2',
+    'fa:16:3e:ed:9a:59': 'foo3',
+    '0c:c4:7a:34:6e:3d': 'oeth1',
+    '0c:c4:7a:34:6e:3c': 'oeth0',
+}
+CFG_DRIVE_FILES_V2 = {
+    'ec2/2009-04-04/meta-data.json': json.dumps(EC2_META),
+    'ec2/2009-04-04/user-data': USER_DATA,
+    'ec2/latest/meta-data.json': json.dumps(EC2_META),
+    'ec2/latest/user-data': USER_DATA,
+    'openstack/2012-08-10/meta_data.json': json.dumps(OSTACK_META),
+    'openstack/2012-08-10/user_data': USER_DATA,
+    'openstack/content/0000': CONTENT_0,
+    'openstack/content/0001': CONTENT_1,
+    'openstack/latest/meta_data.json': json.dumps(OSTACK_META),
+    'openstack/latest/user_data': USER_DATA,
+    'openstack/latest/network_data.json': json.dumps(NETWORK_DATA),
+    'openstack/2015-10-15/meta_data.json': json.dumps(OSTACK_META),
+    'openstack/2015-10-15/user_data': USER_DATA,
+    'openstack/2015-10-15/network_data.json': json.dumps(NETWORK_DATA)}
+
+
+class TestSimpleRun(helpers.FilesystemMockingTestCase):
+    def setUp(self):
+        super(TestSimpleRun, self).setUp()
+        self.new_root = tempfile.mkdtemp()
+        self.tmp = self.tmp_dir()
+        self.sys_cfg = {'datasource': {'ConfigDrive': {'dsmode': 'local'}}}
+        self.paths = cloud_helpers.Paths(
+            {'cloud_dir': self.tmp, 'run_dir': self.tmp})
+        self.ds = ds.DataSourceConfigDrive
+
+    def _patchIn(self, root):
+        self.patchOS(root)
+        self.patchUtils(root)
+
+    @helpers.mock.patch('cloudinit.sources.find_source')
+    def test_invalid_config_drive_subsequent_boot(self, mock_ds):
+        self.addCleanup(shutil.rmtree, self.new_root)
+        self.replicateTestRoot('simple_ubuntu', self.new_root)
+        cfg = {
+            'datasource_list': ['ConfigDrive'],
+            'cloud_init_modules': ['write-files'],
+            'system_info': {'paths': {'run_dir': self.new_root}}
+        }
+        orig_find_devs_with = util.find_devs_with
+        try:
+            # dont' try to lookup for CDs
+            util.find_devs_with = lambda path: []
+            dsrc = self.ds(sys_cfg=self.sys_cfg, distro=None, paths=self.paths)
+            populate_dir(dsrc.seed_dir, CFG_DRIVE_FILES_V2)
+            found = dsrc._get_data()
+            self.assertTrue(found)
+        finally:
+            util.find_devs_with = orig_find_devs_with
```

Comment:
So for config drive, it can be on an iso, but it can also be on a file in the filesystem, /var/lib/cloud/seed/config_drive; 
It may be much easier to populate this directory in your new_root with the config-drive contents;  Then your "reboot without cdrom" just removes the config_drive dir from the seed directory?

Reply:
reworked as per comment